lib/Conversion/DCToHW/DCToHW.cpp

```diff
@@ -0,0 +1,946 @@
+  Value mux(Value index, ValueRange values,
+            std::optional<StringRef> name = {}) {
+    if (values.size() == 2)
+      return b.create<comb::MuxOp>(loc, index, values[1], values[0]);
+
+    return arrayGet(arrayCreate(values), index, name);
+  }
+
+  ///  Muxes a range of values. The select signal is expected to be a 1-hot
+  ///  encoded value.
+  Value ohMux(Value index, ValueRange inputs) {
+    // Confirm the select input can be a one-hot encoding for the inputs.
+    unsigned numInputs = inputs.size();
+    assert(numInputs == index.getType().getIntOrFloatBitWidth() &&
+           "one-hot select can't mux inputs");
+
+    // Start the mux tree with zero value.
+    // Todo: clean up when handshake supports i0.
+    auto dataType = inputs[0].getType();
+    unsigned width =
+        dataType.isa<NoneType>() ? 0 : dataType.getIntOrFloatBitWidth();
+    Value muxValue = constant(width, 0);
+
+    // Iteratively chain together muxes from the high bit to the low bit.
+    for (size_t i = numInputs - 1; i != 0; --i) {
+      Value input = inputs[i];
+      Value selectBit = bit(index, i);
+      muxValue = mux(selectBit, {muxValue, input});
+    }
+
+    return muxValue;
+  }
+
+  OpBuilder &b;
+  Location loc;
+  Value clk, rst;
+  DenseMap<APInt, Value> constants;
+};
+
+/// Creates a Value that has an assigned zero value. For structs, this
+/// corresponds to assigning zero to each element recursively.
+// NOLINTNEXTLINE(misc-no-recursion)
+static Value createZeroDataConst(RTLBuilder &rtlb, Location loc, Type type) {
+  return TypeSwitch<Type, Value>(type)
+      .Case<NoneType>([&](NoneType) { return rtlb.constant(0, 0); })
+      .Case<IntType, IntegerType>([&](auto type) {
+        return rtlb.constant(type.getIntOrFloatBitWidth(), 0);
+      })
+      .Case<hw::StructType>([&](auto structType) {
+        SmallVector<Value> zeroValues;
+        for (auto field : structType.getElements())
+          zeroValues.push_back(createZeroDataConst(rtlb, loc, field.type));
+        return rtlb.b.create<hw::StructCreateOp>(loc, structType, zeroValues);
+      })
+      .Default([&](Type) -> Value {
+        emitError(loc) << "unsupported type for zero value: " << type;
+        assert(false);
+        return {};
+      });
+}
+
+static bool isZeroWidthType(Type type) {
+  if (auto intType = type.dyn_cast<IntegerType>())
+    return intType.getWidth() == 0;
+  return type.isa<NoneType>();
+}
+
+static UnwrappedIO unwrapIO(Location loc, ValueRange operands,
+                            TypeRange results,
+                            ConversionPatternRewriter &rewriter,
+                            BackedgeBuilder &bb) {
+  RTLBuilder rtlb(loc, rewriter);
+  UnwrappedIO unwrapped;
+  for (auto in : operands) {
+    assert(isa<esi::ChannelType>(in.getType()));
+    InputHandshake hs;
+    auto ready = std::make_shared<Backedge>(bb.get(rtlb.b.getI1Type()));
+    auto [data, valid] = rtlb.unwrap(in, *ready);
+    hs.valid = valid;
```

Comment:
nit: You're constructing a few structs like this. Visually, I prefer using a struct initializer with `/*ready=*/` comments. Would allow you to condense 6 lines to one (using `emplace_back`).

FFS C++: WHY DID YOU NOT GET STRUCT DESIGNATED INITIALIZERS UNTIL '20?! You're only 21 years behind C, which has had this feature since C99!